batchflow/sampler.py

```diff
@@ -266,63 +198,174 @@ def truncate(self, high=None, low=None, expr=None, prob=0.5):
+        cumulated = 0
+        samples = []
+        while cumulated < size:
+            # sample points and compute condition-vector
+            sample = self.sampler.sample(size=batch_size)
+            cond = np.ones(shape=batch_size).astype(np.bool)
+            if low is not None:
+                if expr is not None:
+                    cond &= np.greater_equal(expr(sample).reshape(batch_size, -1), low).all(axis=1)
+                else:
+                    cond &= np.greater_equal(sample, low).all(axis=1)
+
+            if high is not None:
+                if expr is not None:
+                    cond &= np.less_equal(expr(sample).reshape(batch_size, -1), high).all(axis=1)
+                else:
+                    cond &= np.less_equal(sample, high).all(axis=1)
+
+            if high is None and low is None:
+                cond &= expr(sample).all(axis=1)
+
+            # check that truncation-prob is not to small
+            _share = np.sum(cond) / batch_size
+            if _share < SMALL_SHARE and ctr > 0:
+                raise ValueError('Probability of region of interest is too small. Try other truncation bounds')
+
+            # get points from region of interest
+            samples.append(sample[cond])
+            cumulated += np.sum(cond)
+            ctr += 1
+
+        return np.concatenate(samples)[:size]
+
+
+class BaseOperationSampler(Sampler):
+    operation = None
+    def __init__(self, left, right, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.left = left
+        self.right = right
+
+    def sample(self, size):
+        if isinstance(self.right, Sampler):
+            return getattr(self.left.sample(size), self.operation)(self.right.sample(size))
+        else:
+            return getattr(self.left.sample(size), self.operation)(np.array(self.right))
+
+
+class AddSampler(BaseOperationSampler):
+    operation = '__add__'
+
+class MulSampler(BaseOperationSampler):
+    operation = '__mul__'
+
+class TruedivSampler(BaseOperationSampler):
+    operation = '__truediv__'
+
+class SubSampler(BaseOperationSampler):
+    operation = '__sub__'
+
+class PowSampler(BaseOperationSampler):
+    operation = '__pow__'
+
+class FloordivSampler(BaseOperationSampler):
+    operation = '__floordiv__'
+
+class ModSampler(BaseOperationSampler):
+    operation = '__mod__'
+
+class RAddSampler(BaseOperationSampler):
+    operation = '__radd__'
+
+class RMulSampler(BaseOperationSampler):
+    operation = '__rmul__'
+
+class RTruedivSampler(BaseOperationSampler):
+    operation = '__rtruediv__'
+
+class RSubSampler(BaseOperationSampler):
+    operation = '__rsub__'
+
+class RPowSampler(BaseOperationSampler):
+    operation = '__rpow__'
+
+class RFloordivSampler(BaseOperationSampler):
+    operation = '__rfloordiv__'
+
+class RModSampler(BaseOperationSampler):
+    operation = '__rmod__'
+
+classes = dict(zip(['__add__', '__mul__', '__truediv__', '__sub__', '__pow__', '__floordiv__', '__mod__',
```

Comment:
This should be at the top of the module and upper-cased

Reply:
Again, to my understanding, this variable can be defined only after all those two-line classes. Which can be defined only after original `Sampler`.